### PR TITLE
[#6] Administrator features

### PR DESCRIPTION
**Pull request by VladimirBudjen**
Users, medicine and pharmacy registration.

### CODE REVIEW

File: src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/service/impl/LogInServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        String userType = user.getClass().getSimpleName();
+        String accessToken = tokenUtils.generateToken(username);
+        int accessExpiresIn = tokenUtils.getAccessTokenExpiresIn();
+        String refreshToken = tokenUtils.generateRefreshToken(username);
+        int refreshExpiresIn = tokenUtils.getRefreshTokenExpiresIn();
+        return new UserTokenState(userType, accessToken, refreshToken, accessExpiresIn, refreshExpiresIn);
+    }
+
+    @Override
+    public void firsLoginPasswordChange(String newPassword) {
```
Comment (lukaDoric) [inline src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/service/impl/LogInServiceImpl.java:53]: Firs :D

File: src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/service/impl/MedicineServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    @Override
+    public List<List<MedicineNameUuidDto>> getAlternativesGroups() {
+        Iterable<Medicine> allMedicine = medicineRepository.findAll();
+        return createAlternativeGroups(allMedicine);
+    }
+
+    private Medicine createBasicMedicine(MedicineDto dto) {
+        Medicine medicine = new Medicine();
```
Comment (lukaDoric) [inline src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/service/impl/MedicineServiceImpl.java:77]: There is so many setters you should do this through MedicineDto constructor.

File: src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/service/impl/PharmacyServiceImpl.java [hunk 3]
```diff
@@ -27,4 +29,14 @@ public PharmacyServiceImpl(PharmacyRepository pharmacyRepository) {
     public Pharmacy findPharmacyByPharmacyAdmin(Long pharmacyAdminId) {
         return pharmacyRepository.findPharmacyByPharmacyAdmin(pharmacyAdminId);
     }
-}
+
+    @Override
+    public Pharmacy save(PharmacyDto dto) {
+        Pharmacy pharmacy = new Pharmacy();
```
Comment (lukaDoric) [inline src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/service/impl/PharmacyServiceImpl.java:35]: Also here, maybe I missing something about setters.

File: src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/service/impl/RegisterPatientServiceImpl.java [hunk 4]
```diff
@@ -64,8 +66,17 @@ public Patient activate(String email, String activationCode) throws BadActivatio
     }
 
     @Override
-    public Patient findByEmail(String email) {
-        return (Patient) userRepository.findByEmail(email);
+    public Patient findByEmail(String email) throws NotAPatientException{
+        User user = userRepository.findByEmail(email);
+        if(user==null){
```
Comment (lukaDoric) [inline src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/service/impl/RegisterPatientServiceImpl.java:71]: ctrl + alt + L

File: src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/helpers/dtoconverters/MedicineConverter.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public interface MedicineConverter {
+    default List<MedicineDto> createResponse(Iterable<Medicine> medicines) {
+        List<MedicineDto> medicinesDto = new ArrayList<>();
+        medicines.forEach(medicine -> {
```
Comment (lukaDoric) [inline src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/helpers/dtoconverters/MedicineConverter.java:15]: There is so many setters you should do this through MedicineDto constructor.

**Review by lukaDoric — CHANGES_REQUESTED** [review]

File: src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/controller/RegisterController.java [hunk 6]
```diff
@@ -8,51 +8,81 @@
...
         } catch (Exception e) {
-            e.printStackTrace();
-            return new ResponseEntity<>("Registration failed!", HttpStatus.INTERNAL_SERVER_ERROR);
+            return new ResponseEntity<>(registrationFailedAlert, HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    @PostMapping("/user")
+    public ResponseEntity<String> registerSystemAdmin(@RequestBody UserRegistrationDto credentials) {
+        User existUser = this.registerUserService.findByEmail(credentials.getEmail());
```
Comment (milicasiriski) [inline src/main/java/rs/ac/uns/ftn/isa/pharmacy/demo/controller/RegisterController.java:66]: Service could return a boolean here

**Review by milicasiriski — CHANGES_REQUESTED** [review]

**Review by lukaDoric — APPROVED** [review]
💯 💯 💯 💯 💯

**Review by milicasiriski — APPROVED** [review]

**Pull request merged**